### PR TITLE
Make checkpointing on train epoch end condition dynamic

### LINKED ISSUE #15180: `ModelCheckpoint` not working when giving unexpected arg to `Trainer`
### Bug description

Model checkpoint not working when `val_check_interval` arg set while `validation_dataloader` given empty. But `val_check_interval` removed from args it works normally. They shouldn't block each other, since checkpoint callback works according to given metric name.

### How to reproduce the bug

```python
callbacks = [ModelCheckpoint(
    monitor="Some/Train/Metric",
    mode="max",
    verbose=True,
    save_last=True,
    save_weights_only=False,
)]

datamodule # Train only datamodule

trainer = pl.Trainer(
    ...
    val_check_interval=0.5,
    ...
)  # Broken

trainer = pl.Trainer(
    ...
    # val_check_interval=0.5,
    ...
)  # Works normal
```

### Error messages and logs

There's no error message.

### Environment

```
* CUDA:
        - GPU:
                - NVIDIA A100-SXM4-40GB
                - NVIDIA A100-SXM4-40GB
        - available:         True
        - version:           11.6
* Lightning:
        - efficientnet-pytorch: 0.7.1
        - pytorch-lightning: 1.7.7
        - pytorch-ranger:    0.1.1
        - segmentation-models-pytorch: 0.3.0
        - torch:             1.12.1+cu116
        - torch-optimizer:   0.3.0
        - torchaudio:        0.12.1+cu116
        - torchmetrics:      0.9.3
        - torchvision:       0.13.1+cu116
* Packages:
        - absl-py:           1.3.0
        - aiohttp:           3.8.3
        - aiosignal:         1.2.0
        - albumentations:    1.3.0
        - alpha-trainer:     2.4.0
        - antlr4-python3-runtime: 4.9.3
        - async-timeout:     4.0.2
        - attrs:             22.1.0
        - cachetools:        5.2.0
        - certifi:           2022.9.24
        - charset-normalizer: 2.1.1
        - checksumdir:       1.2.0
        - click:             8.1.3
        - coloredlogs:       15.0.1
        - colorlog:          6.7.0
        - commonmark:        0.9.1
        - docker-pycreds:    0.4.0
        - efficientnet-pytorch: 0.7.1
        - filelock:          3.8.0
        - flatbuffers:       22.9.24
        - frozenlist:        1.3.1
        - fsspec:            2022.8.2
        - gitdb:             4.0.9
        - gitpython:         3.1.29
        - google-api-core:   2.10.2
        - google-auth:       2.13.0
        - google-auth-oauthlib: 0.4.6
        - google-cloud-core: 2.3.2
        - google-cloud-storage: 2.5.0
        - google-crc32c:     1.5.0
        - google-resumable-media: 2.4.0
        - googleapis-common-protos: 1.56.4
        - grpcio:            1.50.0
        - huggingface-hub:   0.10.1
        - humanfriendly:     10.0
        - hydra-colorlog:    1.2.0
        - hydra-core:        1.2.0
        - idna:              3.4
        - imageio:           2.22.2
        - importlib-metadata: 5.0.0
        - importlib-resources: 5.10.0
        - iterative-stratification: 0.1.7
        - joblib:            1.2.0
        - kornia:            0.6.8
        - markdown:          3.4.1
        - markupsafe:        2.1.1
        - mpmath:            1.2.1
        - multidict:         6.0.2
        - munch:             2.5.0
        - networkx:          2.8.7
        - numpy:             1.23.4
        - oauthlib:          3.2.2
        - omegaconf:         2.2.3
        - onnx:              1.12.0
        - onnxruntime:       1.12.1
        - onnxsim:           0.4.8
        - opencv-python-headless: 4.6.0.66
        - packaging:         21.3
        - pandas:            1.5.0
        - pathtools:         0.1.2
        - pillow:            9.2.0
        - pip:               22.3
        - pkg-resources:     0.0.0
        - pretrainedmodels:  0.7.4
        - promise:           2.3
        - protobuf:          3.19.6
        - psutil:            5.9.3
        - pyasn1:            0.4.8
        - pyasn1-modules:    0.2.8
        - pydeprecate:       0.3.2
        - pygments:          2.13.0
        - pyparsing:         3.0.9
        - python-dateutil:   2.8.2
        - pytorch-lightning: 1.7.7
        - pytorch-ranger:    0.1.1
        - pytz:              2022.5
        - pywavelets:        1.4.1
        - pyyaml:            6.0
        - qudida:            0.0.4
        - regex:             2022.9.13
        - requests:          2.28.1
        - requests-oauthlib: 1.3.1
        - rich:              12.6.0
        - rsa:               4.9
        - scikit-image:      0.19.3
        - scikit-learn:      1.1.2
        - scipy:             1.9.2
        - segmentation-models-pytorch: 0.3.0
        - sentry-sdk:        1.9.10
        - setproctitle:      1.3.2
        - setuptools:        65.5.0
        - shortuuid:         1.0.9
        - six:               1.16.0
        - smmap:             5.0.0
        - sympy:             1.11.1
        - tensorboard:       2.10.1
        - tensorboard-data-server: 0.6.1
        - tensorboard-plugin-wit: 1.8.1
        - threadpoolctl:     3.1.0
        - tifffile:          2022.10.10
        - timm:              0.4.12
        - tokenizers:        0.13.1
        - torch:             1.12.1+cu116
        - torch-optimizer:   0.3.0
        - torchaudio:        0.12.1+cu116
        - torchmetrics:      0.9.3
        - torchvision:       0.13.1+cu116
        - tqdm:              4.64.1
        - transformers:      4.23.1
        - typing-extensions: 4.4.0
        - urllib3:           1.26.12
        - wandb:             0.13.4
        - werkzeug:          2.2.2
        - wheel:             0.37.1
        - yarl:              1.8.1
        - zipp:              3.9.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #24~20.04.1-Ubuntu SMP Mon Sep 12 06:14:01 UTC 2022
```

### More info

_No response_

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #15180

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj

### CODE REVIEW

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -262,13 +262,6 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: s
         if trainer.is_global_zero and stage == "fit":
             self.__warn_if_dir_not_empty(self.dirpath)
 
-        # NOTE: setting these attributes needs to happen as early as possible BEFORE reloading callback states,
-        # because the attributes are part of the state_key which needs to be fully defined before reloading.
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:266]: What about this comment? This requirement no longer holds if I understand correctly, because we evaluate the condition `self._should_save_on_train_epoch_end(trainer)` on the fly and hence later than requested by this note.

This is the reason why the existing tests had to be changed, right?
Can you describe the consequences this has on existing workflows?

Comment (carmocca) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:266]: Good point. Yeah, this blocks making this attribute dynamic, as the callback state depends on its value.

I don't think we can proceed with this proposal

Comment (rohitgr7) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:266]: > Can you describe the consequences this has on existing workflows?

for existing workflows, it will have a boolean value, so the dynamic behavior won't do anything.
```py
if self._save_on_train_epoch_end is not None:
    return self._save_on_train_epoch_end
```

Comment (carmocca) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:266]: I had a discussion with Rohit in a call.

This is okay. There's a breaking change in that existing checkpoints might not get loaded into the callback instance because the instance will have `None` but the checkpoint data `bool`. I don't think we can do anything about it since there's no way for us to "upgrade" the checkpoint. We cannot differentiate between an user set value and an automatically set value from just the checkpoint.

An alternative would be to remove `save_on_train_epoch_end` from the state key, but that would also be a breaking change

File: src/pytorch_lightning/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,11 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ### Changed
 
 
+### Fixed
+
+- Made checkpointing on train epoch end condition dynamic ([#15300](https://github.com/Lightning-AI/lightning/pull/15300))
```
Comment (carmocca) [inline src/pytorch_lightning/CHANGELOG.md:17]: I would consider this a "feature" if we make it dynamic. If you want to fix the linked issue in a bug-fix release, then there should be 2 PRs, one to fix the static issue, and another to add dynamic support

**Review by carmocca — APPROVED** [review]

File: tests/tests_pytorch/checkpointing/test_model_checkpoint.py [hunk 3]
```diff
@@ -1360,3 +1360,9 @@ def test_save_last_every_n_epochs_interaction(tmpdir, every_n_epochs):
         trainer.fit(model)
     assert mc.last_model_path  # a "last" ckpt was saved
     assert save_mock.call_count == trainer.max_epochs
+
+
+def test_train_epoch_end_ckpt_with_no_validation():
+    trainer = Trainer(val_check_interval=0.5)
+    trainer.num_val_batches = [0]
+    assert trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
```
Comment (awaelchli) [inline tests/tests_pytorch/checkpointing/test_model_checkpoint.py:1368]: To verify the dynamic behavior
```suggestion
    trainer.num_val_batches = [0]
    assert trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
    trainer.num_val_batches = [1]
    assert not trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
    trainer.val_check_interval = 0.8
    assert not trainer.checkpoint_callback._should_save_on_train_epoch_end(trainer)
```

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
